toxcore/group.c

```diff
@@ -217,661 +294,1009 @@ static uint64_t calculate_comp_value(const uint8_t *pk1, const uint8_t *pk2)
-            g->close[i].type = GROUPCHAT_CLOSE_NONE;
-            kill_friend_connection(g_c->fr_c, friendcon_id);
-            return 0;
+                            g->peers_list[g->numpeers_in_list] = (uint16_t)i;
+                            ++g->numpeers_in_list;
+                            some_changes = true;
+                        }
+                    }
+                }
+            }
         }
-    }
 
-    return -1;
-}
+        if (g_c->group_namelistchange) {
+            if (some_changes) {
 
+                if (!g->invite_called && !g->join_mode) {
+                    if (g_c->invite_callback) {
 
-/*
- * Delete a peer from the group chat.
- *
- * return 0 if success
- * return -1 if error.
- */
-static int delpeer(Group_Chats *g_c, int groupnumber, int peer_index, void *userdata)
-{
-    Group_c *g = get_group_c(g_c, groupnumber);
+                        g->fake_join = true;
+                        g_c->invite_callback(g_c->m, UINT32_MAX, g->identifier[0], g->identifier + 1, GROUP_IDENTIFIER_LENGTH - 1, userdata);
+                        g->fake_join = false;
 
-    if (!g) {
-        return -1;
-    }
+                        if (!g->invite_called) {
+                            del_groupchat(g_c, (int)groupnumber);
+                            return;
+                        }
+                    }
+                }
 
-    uint32_t i;
+                g_c->group_namelistchange(g_c->m, (int)groupnumber, 0, CHAT_CHANGE_OCCURRED, userdata);
+                g->nick_changed = false;
+            }
+        }
 
-    for (i = 0; i < DESIRED_CLOSE_CONNECTIONS; ++i) { /* If peer is in closest_peers list, remove it. */
-        if (g->closest_peers[i].entry && id_equal(g->closest_peers[i].real_pk, g->group[peer_index].real_pk)) {
-            g->closest_peers[i].entry = 0;
-            g->changed = GROUPCHAT_CLOSEST_REMOVED;
-            break;
+        /* and now rebuild closest */
+
+        uint16_t old_closest_peers[DESIRED_CLOSE_CONNECTIONS];
+        size_t inclose = 0;
+
+        for (i = 0; i < DESIRED_CLOSE_CONNECTIONS; ++i) {
+            if (CLOSEST(i)) {
+                old_closest_peers[inclose++] = g->closest_peers[i];
+            }
         }
-    }
 
-    int friendcon_id = getfriend_conn_id_pk(g_c->fr_c, g->group[peer_index].real_pk);
+        g->closest_peers_entry = 0;
 
-    if (friendcon_id != -1) {
-        remove_close_conn(g_c, groupnumber, friendcon_id);
-    }
+        size_t np = g->numpeers;
 
-    --g->numpeers;
+        for (i = 0; i < np; ++i) {
 
-    void *peer_object = g->group[peer_index].object;
+            Group_Peer *peer = g->peers + i;
 
-    if (g->numpeers == 0) {
-        free(g->group);
-        g->group = NULL;
-    } else {
-        if (g->numpeers != (uint32_t)peer_index) {
-            memcpy(&g->group[peer_index], &g->group[g->numpeers], sizeof(Group_Peer));
+            if (peer->friendcon_id == -2 || peer->auto_join || id_equal(g->real_pk, peer->real_pk)) {
```

Comment:
Looks like -2 means something. When you add/use a constant, make sure it's used throughout the code.